design/nodes-machines-and-hosts.md

```diff
@@ -0,0 +1,213 @@
+
+`BareMetalHost` objects represents a physical computer, including its
+hardware inventory and information to access the onboard management
+controller. The BareMetalHost (or "host") CRD is defined in the
+[baremetal-operator](https://github.com/metalkube/baremetal-operator/)
+from the MetalKube project. The CRD is made up of completely custom
+data, needed to fully realize a Kubernetes-native API for a
+self-hosted baremetal provisioning tool. The lifecycle of a host is
+based on the availability of hardware, rather than its use as part of
+the cluster.
+
+#### Growing the Cluster
+
+The admin user interacts with MetalKube by first registering hosts to
+be part of the inventory available to the provisioning system. Each
+host is then matched against known hardware profiles (comparing
+information such as NICs, storage, CPUs, RAM, etc.). Hosts matching
+known profiles are marked as being ready to be used. The admin (or
+autoscaler) can then create a new Machine object as a request to grow
+the cluster.
+
+The actuator in the
+[cluster-api-provider-baremetal](https://github.com/metalkube/cluster-api-provider-baremetal)
+will match the hardware profile in the provider spec of the Machine
+object to available hosts with the same profile. When an available
+host is found, the actuator links it to the Machine, and adds
+information to the host to tell the baremetal-operator which image to
+provision to it, which machine configuration to use, etc. After the
+host is provisioned and rebooted, kubelet connects to the cluster in
+the usual way, triggering the creation of a Node object.
+
+#### Shrinking the Cluster
+
+The admin user (or autoscaler) can similarly delete the Machine
+object, breaking the link to the host. When that happens the host is
+completely evacuated (going through the standard Kubernetes drain
+procedure, as well as removing any storage replicas it contains). The
+host is then powered off and placed back into the available inventory
+pool.
+
+To completely remove a host, the BareMetalHost object is
+deleted. Whether this wipes the storage in the host is yet to be
+determined.
+
+### Risks and Mitigations
+
+We risk some confusion by introducing a third entity tied to the
+"computer" that is part of the cluster. For other providers, the
+information we are storing in the host record would be stored in some
+external database (typically behind a cloud API). There will be a few
+new operations that admin users will perform on hosts that are not
+typically performed on compute resources managed by other providers;
+powering a host off to perform hardware maintenance is the main such
+use case.
+
+## Design Details
+
+![Component Relationships](component-relationships.png "Component Relationships")
+
+### Work Items
+
+* Write and maintain this document
+
+### Dependencies
+
+N/A
+
+### Test Plan
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+N/A
+
+### Version Skew Strategy
+
+N/A
+
+## Drawbacks [optional]
+
+N/A
+
+## Alternatives
```

Comment:
Should we add a second alternative discussion covering what it would take to make the Machine object work for us?

To me, the key thing is that the Machine needs to be decoupled from Kubernetes.  A valid Machine would need to be able to represent a host that is intentionally off, or intentionally not being provisioned because of some policy reason (hardware profiles, the current desired size of the cluster, as examples).

We can link to this discussion: https://github.com/kubernetes-sigs/cluster-api/issues/721

That describes a direction that makes it less necessary to have BareMetalHost, but it also doesn't look like it's happening any time soon, and we're looking to implement this ASAP.

We should acknowledge that once we create this, it will be very, very hard to drop it.  We could further acknowledge that at some future point, the separate CRD may make less sense depending on the direction of the cluster-api project, but I also don't imagine it will be harmful to have it.

Reply:
> Should we add a second alternative discussion covering what it would take to make the Machine object work for us?

Considering a lot of the feedback on the UX side so far, I'd love to see something here if it's possible. Even if we consider the designs towards something like this being future future.

Reply:
I added this. I'm not sure it would actually be that hard to deal with an upgrade. The new controller could just import the existing host objects into Machines and then delete them, right? I described the case without getting into the upgrade question.